src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs

```diff
@@ -0,0 +1,1202 @@
+    {
+        static void Main()
+        {
+        }
+
+        static ref int Test1(S1 arg)
+        {
+            return ref (new int[1])[0];
+        }
+
+        static S1 MayWrap(ref int arg)
+        {
+            return default;
+        }
+
+        static ref int Test3()
+        {
+            int local = 42;
+            var sp = MayWrap(ref local);
+            return ref Test1(sp);
+        }
+
+        ref struct S1
+        {
+        }
+    }
+";
+            CreateStandardCompilation(text).VerifyDiagnostics(
+                // (22,30): error CS8520: Cannot use local 'sp' in this context because it may expose referenced variables outside of their declaration scope 
+                //             return ref Test1(sp);
+                Diagnostic(ErrorCode.ERR_EscapeLocal, "sp").WithArguments("sp").WithLocation(22, 30),
+                // (22,24): error CS8521: Cannot use a result of 'Program.Test1(Program.S1)' in this context because it may expose variables referenced by parameter 'arg' outside of their declaration scope
+                //             return ref Test1(sp);
+                Diagnostic(ErrorCode.ERR_EscapeCall, "Test1(sp)").WithArguments("Program.Test1(Program.S1)", "arg").WithLocation(22, 24)
+            );
+        }
+
+        [Fact()]
+        public void RefLikeReturnEscapeWithRefLikes()
+        {
+            var text = @"
+    class Program
+    {
+        static void Main()
+        {
+        }
+
+        static ref int Test1(ref S1 arg)
+        {
+            return ref (new int[1])[0];
+        }
+
+        static S1 MayWrap(ref int arg)
+        {
+            return default;
+        }
+
+        static ref int Test3()
+        {
+            int local = 42;
+            var sp = MayWrap(ref local);
+            return ref Test1(ref sp);    // error
+        }
+
+        static ref int Test4()
+        {
+            var sp = MayWrap(ref (new int[1])[0]);
+
+            // valid. 
+            // Even though sp is itself not ref-returnable, Test1 cannot ref-return it
+            return ref Test1(ref sp);    // OK
+        }
+
+        ref struct S1
+        {
+        }
+    }
+";
+            CreateStandardCompilation(text).VerifyDiagnostics(
+                // (22,34): error CS8520: Cannot use local 'sp' in this context because it may expose referenced variables outside of their declaration scope 
+                //             return ref Test1(ref sp);    // error
+                Diagnostic(ErrorCode.ERR_EscapeLocal, "sp").WithArguments("sp").WithLocation(22, 34),
+                // (22,24): error CS8521: Cannot use a result of 'Program.Test1(ref Program.S1)' in this context because it may expose variables referenced by parameter 'arg' outside of their declaration scope
+                //             return ref Test1(ref sp);    // error
+                Diagnostic(ErrorCode.ERR_EscapeCall, "Test1(ref sp)").WithArguments("Program.Test1(ref Program.S1)", "arg").WithLocation(22, 24)
```

Comment:
Should we try to suppress the second diagnostic? Seems cascading.

Reply:
That is intentional. The idea is that it might be hard to figure the actual reason for an escape failure, so we produce intermediate errors. If this is not helpful we can try something else. 

---
In reply to: [135401643](https://github.com/dotnet/roslyn/pull/21334#discussion_r135401643) [](ancestors = 135401643)